### DISCUSSION TITLE
Line/column numbers from `Located`/`Span`

### DISCUSSION THREAD

**Discussion by 9999years** [winnow-rs/winnow#330] [Q&A]
Hello! I've read the [`winnow` introduction post](https://epage.github.io/blog/2023/02/winnow-toml-edit-combine-nom/), which talks about the design of `Located` in `winnow` compared to `nom-locate`. However, I'm not sure how exactly to get line/column numbers from spans.

The [`line_numbers` crate](https://docs.rs/line-numbers/latest/line_numbers/struct.LinePositions.html) can convert from `usize` offsets to line numbers (although it doesn't help with column numbers, even byte-based ones).

I would assume this is a fairly common use-case, so I'm not sure if there's a suggested workflow here.

For my specific use-case, I'm parsing through a file and extracting some spans (commands to evaluate like doctests). Then, my program will evaluate those commands and report the results to the user along with where the span was defined (file, line number, ideally column number).

Using [`Parser::with_span`](https://docs.rs/winnow/latest/winnow/trait.Parser.html#method.with_span) I can extract `usize` offsets and later convert them to line numbers, but this strikes me as potentially inefficient: each `usize`-to-line-number conversion requires an O(n) iteration through the codepoints in the file contents, so it would be nice to be able to at least convert them all at once, and perhaps ideally to have them tracked as parsing happens. However, I recall that the overhead from some of that tracking hurt performance substantially.

In the meantime, I can implement a naïve approach with `line_numbers`, but I'm curious if there's a better way to do this.

**Comment by 9999years** [comment]
To convert multiple `usize` offsets to line numbers at once, I think I'm going to use the [`line_span` crate](https://docs.rs/line-span/latest/line_span/).

**Comment by epage** [comment]
>  so it would be nice to be able to at least convert them all at once, and perhaps ideally to have them tracked as parsing happens. However, I recall that the overhead from some of that tracking hurt performance substantially.

Yes, there are two problems
- Line/column is presentation information and so isn't essential for the parser while tracking it proactively can negative impact performance
- Providing column information either requires domain knowledge (whether unicode grapheme clusters should be used, etc) or giving bytes which will require post-processing by the caller anyways

Glad you found that crate and it works well for you.
